### PR TITLE
add sns_lifecycle.yml

### PR DESCRIPTION
**Pull request by mraszyk**
This PR builds the sns-testing Docker image and runs the basic scenario in it. It also adds a few Bash scripts for awaiting NNS/SNS proposals and canisters to be in running state.

### CODE REVIEW

File: .github/workflows/sns_lifecycle.yml [hunk 1]
```diff
@@ -0,0 +1,101 @@
+name: "SNS lifecycle"
+on:
+  push:
+jobs:
+  sns_lifecycle:
+    runs-on: ubuntu-latest
```
Comment (aterga) [inline .github/workflows/sns_lifecycle.yml:6]: We should probably opt for a deterministic pipeline, so I would fix the versions of all the dependencies.

**Comment by aterga** [comment]
Thanks for the PR, @mraszyk !

IIUC, this pipeline ensures that there is a version of Ubuntu on which the sns-testing scripts work. However, directly running on Ubuntu is not one of the recommended workflows described in the README.md file.

I think if we decide that sns-testing CI is important, we should build a pipeline that runs the Docker-based solution. This should be rather straightforward, as one would not need to explicitly install any dependencies. So typically a docker-based solution has a CI pipeline that:

1. Builds a docker image from the latest commit.
2. Starts the Docker image and runs some tests.
3. If we're on the main branch and the tests succeed, it then publishes the image to Dockerhub.

File: .github/workflows/sns_lifecycle.yml (hunk 1 above)
Comment (mraszyk) [reply .github/workflows/sns_lifecycle.yml]: Pinned ubuntu to ubuntu-22.04.

File: README.md [hunk 2]
```diff
@@ -132,7 +132,7 @@ After getting familiar with the basic scenario, you may replace the test caniste
 2. Start a local replica instance:
     ```bash
    SNS_TESTING_INSTANCE=$(
-       docker run -p 8080:8080 -v "`pwd`":/dapp -d martin2718/sns-testing:latest dfx start --clean
+       docker run -p 8080:8080 -v "`pwd`":/dapp -d dfinity/sns-testing:master dfx start --clean
```
[the remark this answers is not in the document]
Comment (mraszyk) [reply README.md]: > Are we already publishing this image under the dfinity account?

This is what this PR is supposed to achieve.

File: .github/workflows/sns_lifecycle.yml [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+    needs: sns_lifecycle
+    runs-on: ubuntu-22.04
+    steps:
+    - name: Checkout code
+      uses: actions/checkout@v2
+    - name: Log in to Docker Hub
+      uses: docker/login-action@f4ef78c080cd8ba55a85445d5b36e214a81df20a
+      with:
+        username: ${{ secrets.DOCKER_USERNAME }}
+        password: ${{ secrets.DOCKER_PASSWORD }}
```
Comment (aterga) [inline .github/workflows/sns_lifecycle.yml:54]: Who will effectively administer these secrets? IDX?

Comment (mraszyk) [reply .github/workflows/sns_lifecycle.yml:54]: I've reached out to IT about that.

File: README.md [hunk 4]
```diff
@@ -132,7 +132,7 @@ After getting familiar with the basic scenario, you may replace the test caniste
 2. Start a local replica instance:
     ```bash
    SNS_TESTING_INSTANCE=$(
-       docker run -p 8080:8080 -v "`pwd`":/dapp -d martin2718/sns-testing:latest dfx start --clean
+       docker run -p 8080:8080 -v "`pwd`":/dapp -d dfinity/sns-testing:main dfx start --clean
```
Comment (aterga) [inline README.md:135]: What's the meaning of the tag `:main`? Typically, images have `:latest`, `:<version-number>`, or no tags at all (if they were built locally).

Comment (mraszyk) [reply README.md:135]: This is the tag we'll get using the default [template](https://docs.docker.com/build/ci/github-actions/manage-tags-labels/) for managing tags and labels with Github actions (also see [here](https://github.com/docker/metadata-action#basic))

Comment (aterga) [reply README.md:135]: Sounds reasonable

**Comment by aterga** [comment]
Could we factor out the Docker publishing step from this PR into a separate one?

I think we could already merge this (modulo logging into the Docker account), and it would be beneficial for us to do that ASAP. Wdyt @mraszyk ?

**Comment by mraszyk** [comment]
> Could we factor out the Docker publishing step from this PR into a separate one?

Indeed. Done.

File: .github/workflows/sns_lifecycle.yml (hunk 3 above)
Comment (mraszyk) [reply .github/workflows/sns_lifecycle.yml]: Factored out from this PR.

**Review by aterga — APPROVED** [review]

**Pull request merged**